File: src/main/java/linkedList/postorder/reverseBetween.py
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

#
# 
# @param head ListNode类 
# @param m int整型 
# @param n int整型 
# @return ListNode类
#


class Solution:

    def __init__(self):
        self.nxt = None

    def reverseBetween(self, head, m, n):
        # write code here
        if m == 1:
            return self.reverseN(head, n)
        head.next = self.reverseBetween(head.next, m - 1, n - 1)
        return head

    def reverseN(self, head, n):
        if n == 1:
            self.nxt = head.next
            return head
        newHead = self.reverseN(head.next, n - 1)
        head.next.next = head
        head.next = self.nxt
        return newHead

File: src/main/java/linkedList/postorder/reverse_between.go
package main

func main() {
	test := &ListNode{Val: 1}
	test.Next = &ListNode{Val: 3}
	test.Next.Next = &ListNode{Val: 2}
}

func reverseBetween(head *ListNode, left int, right int) *ListNode {
	if left == 1 {
		return reverseN(head, right)
	}
	head.Next = reverseBetween(head.Next, left-1, right-1)
	return head
}

var nxt *ListNode

func reverseN(head *ListNode, right int) *ListNode {
	if right == 1 {
		nxt = head.Next
		return head
	}
	newHead := reverseN(head.Next, right-1)
	head.Next.Next = head
	head.Next = nxt
	return newHead
}
